src/conduits/RwaOutputConduit3.t.sol

```diff
@@ -0,0 +1,630 @@
+        outputConduit.push(500 ether);
+
+        assertEq(usdx.balanceOf(address(me)), 500 * USDX_BASE_UNIT);
+        assertEq(usdx.balanceOf(address(outputConduit)), 100 * USDX_BASE_UNIT);
+        assertApproxEqAbs(dai.balanceOf(address(outputConduit)), 0, USDX_DAI_DIF_DECIMALS);
+        assertEq(outputConduit.to(), address(0));
+    }
+
+    function testPushAmountWhenAlreadyHaveSomeGemBalanceGetAll() public {
+        usdx.mint(100 * USDX_BASE_UNIT);
+        usdx.transfer(address(outputConduit), 100 * USDX_BASE_UNIT);
+
+        assertEq(outputConduit.to(), me);
+        assertEq(usdx.balanceOf(me), 0);
+        assertEq(usdx.balanceOf(address(outputConduit)), 100 * USDX_BASE_UNIT);
+        assertEq(dai.balanceOf(address(me)), 1000 ether);
+
+        dai.transfer(address(outputConduit), 500 ether);
+
+        assertEq(dai.balanceOf(me), 500 ether);
+        assertEq(dai.balanceOf(address(outputConduit)), 500 ether);
+
+        vm.expectEmit(true, true, false, false);
+        emit Push(address(me), 500 * USDX_BASE_UNIT);
+        outputConduit.push();
+
+        assertEq(usdx.balanceOf(address(me)), 500 * USDX_BASE_UNIT);
+        assertEq(usdx.balanceOf(address(outputConduit)), 100 * USDX_BASE_UNIT);
+        assertEq(outputConduit.to(), address(0));
+    }
+
+    function testPushAmountWhenAlreadyHaveSomeDaiBalance() public {
+        assertEq(outputConduit.to(), me);
+        assertEq(usdx.balanceOf(me), 0);
+        assertEq(dai.balanceOf(address(me)), 1000 ether);
+
+        dai.transfer(address(outputConduit), 500 ether);
+
+        assertEq(dai.balanceOf(me), 500 ether);
+        assertEq(dai.balanceOf(address(outputConduit)), 500 ether);
+
+        vm.expectEmit(true, true, false, false);
+        emit Push(address(me), 400 * USDX_BASE_UNIT);
+        // push ionly 400 leave 100 dai in conduit
+        outputConduit.push(400 ether);
+
+        assertEq(usdx.balanceOf(address(me)), 400 * USDX_BASE_UNIT);
+        assertApproxEqAbs(dai.balanceOf(address(outputConduit)), 100 ether, USDX_DAI_DIF_DECIMALS);
+        assertEq(outputConduit.to(), address(0));
+    }
+
+    function testPushAmountFuzz(uint256 wad) public {
+        assertEq(outputConduit.to(), me);
+        dai.transfer(address(outputConduit), dai.balanceOf(me));
+        uint256 cDaiBalance = dai.balanceOf(address(outputConduit));
+        uint256 usdxBalance = usdx.balanceOf(me);
+
+        wad = bound(wad, 10**18, cDaiBalance);
+
+        vm.expectEmit(true, true, false, false);
+        emit Push(address(me), wad / USDX_DAI_DIF_DECIMALS);
+        outputConduit.push(wad);
+
+        assertEq(usdx.balanceOf(me), usdxBalance + wad / USDX_DAI_DIF_DECIMALS);
+        // We lose some dust because of decimals diif dai.decimals() > gem.decimals()
+        assertApproxEqAbs(dai.balanceOf(address(outputConduit)), cDaiBalance - wad, USDX_DAI_DIF_DECIMALS);
+        assertEq(outputConduit.to(), address(0));
+    }
+
+    function testExpectedGemAmountFuzz(uint256 wad) public {
```

Comment:
mmm we should probably test the helper functions with greater than zero `tin` and `tout`, otherwise they don't do much (except decimals change).

This might be covered in other tests (is it?), but in that case I'd say fuzzing here just for the decimals change is probably an overkill.

Reply:
Done

Reply:
I see you added non zero fees for `testRequiredGemAmountFuzz` and `testExpectedGemAmountFuzz`, but not for `testRequiredDaiAmountFuzz`.

Also, we now have 4 helpers, which are pretty symmetrical, but seems like we don't have a (fuzz) test for `expectedDaiWad`.
I'd add that as we have the other 3.

Reply:
Add PSM fee for `testRequiredDaiAmountFuzz ` and add missing tests for `expectedDaiWad `.